docs/cluster_api.md

```diff
@@ -0,0 +1,95 @@
+# Cluster Operator API
+
+## Overview
+
+The cluster operator API defines a set of resources that allow the instantiation of OpenShift clusters and subsequent
+updates to those clusters based on changes to said resources.
+
+The general approach for these resources is to follow the Kubernetes model of a spec section that is end-user controlled and
+a status section that represents the current state of the resources. Changes to the spec will result in the necessary actions
+by the system to reconcile the current state of the system with what the user has specified.
+
+## Resources
+
+### Cluster
+
+The cluster resource should allow 2 levels/scopes of spec:
+- Cluster-wide configuration
+- NodeGroup configuration
+
+Furthermore, each scope can be described in 2 parts:
+- Hardware spec (needed to instantiate the proper virtual hardware)
+- OpenShift config (openshift-specific configuration)
+
+For the Status part, if the cluster has been instantiated, we need to know how to access it (Kubeconfig may be sufficient)
+
+In more detail:
+
+#### Spec:
+* Cluster-wide configuration:
+  * ObjectMetadata (inline ObjectMetadata)
+  * TypeMetadata (inline TypeMetadata)
+  * Hardware-spec
+    * Type (AWS, Azure, GCE, etc) [string]
+    * AWSConfig: [struct]
+      - AWSSecret [LocalReference]
+      - Region [string]
+      - VPCName [string]
+      - VPCSubnet [string]
+  * OpenShift Configuration:
+    * DeploymentType [string]
+    * OpenShiftVersion [string]
```

Comment:
Will reving the OpenShiftVersion field (ie 3.6 to 3.7) cause the nodegroups to point to a new AMI? Or will an upgrade simply be updating the AMI on the nodegroup(s).

Reply:
My understanding is that this is more for the use of the master. For nodes themselves, yes we would just change the AMI

Reply:
This is along the lines of the AMI comment, but the code you deploy with needs to be tied together.  There are two cases:

1. use binaries, images, and AMIs that come from a concrete release tag
2. use binaries, images, and AMIs that come from arbitrary locations (PR jobs, for example)

The API consumer needs to be able to say "upgrade to v3.10.1" but the test infrastructure needs to be able to say "upgrade to v3.11.0-prerelease.pr-19034 which comes from here, here, and here".  That's probably calling out that you may be missing another construct like a "Version" (which contains AMI names, image names, etc) and some API callers can override those on the field.

Reply:
What is 'deployment type' ?
Do we really want to call this out as a 'master config'?
Could we maybe declare 'containerized only' for the master components? Then 'version' is really 'image name'. Which should let us write higher level controllers to do anything smart around upgrading. It also starts to treat masters a bit like we're eventually going to treat the router/registry/et al.

Reply:
Deployment type per openshift ansible is essentially "origin" or "openshift-enterprise". (I believe that is what was intended here)

Containerized masters would be interesting, makes me wonder if we should target Atomic as well, and where things are with the recent discussion around system containers + static pods/daemonsets. I'll hunt down some info on what we should target for this.

Reply:
This also got me thinking about the conversation awhile back on moving towards system containers for the essentials (kubelet and container runtime) and then static pods or daemon sets for everything else. Checked in with Scott this morning and it sounds like that's still the plan for the 3.10 or 3.11 timeframe so we should keep that containers first approach in mind, even though the golden images today are entirely RPM based.

Upstream cluster API already has a version struct differentiating kubelet/control plane/docker version, we'll need to think about the same for OpenShift and cri-o.

Reply:
> What is 'deployment type'

it's this variable: https://github.com/openshift/openshift-ansible/blob/master/inventory/hosts.example#L30-L31

Reply:
Deployment type is a legacy of a stupid choice we had to make long ago to split our packages.  We really need to collapse all of that - install should care about only:

1. where it gets RPMs
2. what is the matching golden image for that AMI
3. what is the location of images to fetch

We should be collapsing down everything else behind those.  My point about version is that those three are tied - you will always get them as a group.  An end user facing value is "v3.7.9".  A machine facing value is "v3.7.9-pr19384-x-93842948" - both are use cases that the API needs to tolerate.  I would expect that the locations of binaries and the image name would be part of a Version resource and referenced.  I'm not opposed to also supporting direct reference.